### ISSUE TITLE
ghc fails sanity check on OS X with Nix

### ISSUE THREAD

**Issue by usharf** [commercialhaskell/stack#960]
Similar to  #497, but on OS X with Nix, I'm getting "The GHC located at /Users/uri/.nix-profile/bin/ghc failed to compile a sanity check.", though no problem to use cabal or `nix-shell`. Stack won't setup or build.

My setup: 

```sh
▶  which stack
/Users/uri/.nix-profile/bin/stack

▶ stack --version
Version 0.1.4.1 (UNKNOWN commits) X86_64
```
```sh
▶ env | grep NIX
NIXDIR=/Users/uri/nixpkgs
NIX_PATH=nixpkgs=/Users/uri/nixpkgs
NIX_USER_PROFILE_DIR=/nix/var/nix/profiles/per-user/uri
NIX_GHC=/nix/store/hxmb7pcax73ygw9k58yb4y2zxva57ds6-ghc-7.10.2/bin/ghc
NIX_GHCPKG=/nix/store/hxmb7pcax73ygw9k58yb4y2zxva57ds6-ghc-7.10.2/bin/ghc-pkg
NIX_GHC_DOCDIR=/nix/store/hxmb7pcax73ygw9k58yb4y2zxva57ds6-ghc-7.10.2/share/doc/ghc/html
NIX_GHC_LIBDIR=/nix/store/hxmb7pcax73ygw9k58yb4y2zxva57ds6-ghc-7.10.2/lib/ghc-7.10.2
```

Same error when I remove the `NIX_GHC*` environment setup above (which is what the  `~/.nix-profile/bin/ghc` wrapper sets up as well.

When I switch into a `nix-shell` though, it seem to work.
```sh
[nix-shell:~/Projects/haskell/bassbull]$ stack setup
stack will use the GHC on your PATH
For more information on paths, see 'stack path' and 'stack exec env'
To use this GHC and packages outside of a project, consider using:
stack ghc, stack ghci, stack runghc, or stack exec
```

**Comment by snoyberg** [comment]
Can you try compiling the following program outside of the Nix shell?

```haskell
import Distribution.Simple
main = putStrLn "Hello World"
```

My guess is that there's a GHC on your path that is not properly configured to run outside of the nix shell, and that you should either:

1. Remove it from your PATH
2. Tell stack not to use it via `--no-system-ghc` (or `system-ghc: false` in stack.yaml).

**Comment by rvion** [comment]
some ideas:
 1. `--no-system-ghc` being a default ?
 2. Printing a warning message when system ghc is used 
 3. Preferring stakc ghc binary over system ghc when possible

**Comment by usharf** [comment]
@snoyberg yes, I do. I thought running within stack should isolate me...
Regardless, it fails with similar error to what I see if I try to `stack ghci` etc.

```sh
▶ ghc --make Main.hs
[1 of 1] Compiling Main             ( Main.hs, Main.o ) [flags changed]
Linking Main ...

In file included from /var/folders/c0/4fzwf9wx4qvdj10z6c52_s_m0000gn/T/ghc77217_0/ghc_4.c:1:0:
    

In file included from /nix/store/zpf65w2k132d1x1wi4ynlznj2rljx904-ghc-7.10.2/lib/ghc-7.10.2/include/Rts.h:30:0:
    

/nix/store/zpf65w2k132d1x1wi4ynlznj2rljx904-ghc-7.10.2/lib/ghc-7.10.2/include/Stg.h:69:10:
     fatal error: 'math.h' file not found
#include <math.h>
         ^
1 error generated.
```

Not it works fine as suggested with `--no-system-ghc`. 
System ghc seems to still fail though.

**Comment by snoyberg** [comment]
Barring any change in defaults (which I don't think we're going to do), it sounds like stack is doing the right thing here. Can you recommend any changes that would avoid confusion here?

**Comment by usharf** [comment]
IMHO thought the full message when `stack setup` does indicate it's attempting to use system `ghc, it's easy to miss the point, if you're not sure what you are looking at/for... 

also what are the advantage of stack on nix compared to nix-shell, or are they mutually exclusive?

```sh
▶ stack setup
The GHC located at /Users/uri/.nix-profile/bin/ghc failed to compile a sanity check. Please see:

    https://github.com/commercialhaskell/stack/wiki/Downloads

for more information. Exception was:
Running /Users/uri/.nix-profile/bin/ghc /var/folders/c0/4fzwf9wx4qvdj10z6c52_s_m0000gn/T/stack-sanity-check81989/Main.hs -no-user-package-db exited with ExitFailure 1
[1 of 1] Compiling Main             ( /var/folders/c0/4fzwf9wx4qvdj10z6c52_s_m0000gn/T/stack-sanity-check81989/Main.hs, /var/folders/c0/4fzwf9wx4qvdj10z6c52_s_m0000gn/T/stack-sanity-check81989/Main.o )
Linking /var/folders/c0/4fzwf9wx4qvdj10z6c52_s_m0000gn/T/stack-sanity-check81989/Main ...

In file included from /var/folders/c0/4fzwf9wx4qvdj10z6c52_s_m0000gn/T/ghc81993_0/ghc_4.c:1:0:
    

In file included from /nix/store/zpf65w2k132d1x1wi4ynlznj2rljx904-ghc-7.10.2/lib/ghc-7.10.2/include/Rts.h:30:0:
```

**Comment by snoyberg** [comment]
I didn't understand your first paragraph, could you rephrase?

Regarding the difference between stack and nix, the mailing list or stack
overflow would be a better place for such a question, as there are people
using both that can answer that.

**Comment by usharf** [comment]
Sorry, Was trying to say that it is not obvious why the installation is failing.
Thanks. I'm closing this issue.

**Comment by nrolland** [comment]
I would expect any stack command to preferentially use a stack ghc as suggested by @rvion
